### DISCUSSION TITLE
[5일차 - 예진] 실행 컨텍스트

### DISCUSSION THREAD

**Discussion by ez1n** [zzangCS/zzangInterview#19] [프론트 면접 질문]
실행 컨텍스트에 대해 설명해주세요.

**Comment by ez1n** [comment]
**실행 컨텍스트**

> 코드가 실행되기 위해 필요한 정보들을 담고 있는 곳
Scope, Hoisting, thisk function, clouse 등의 동작 원리를 담고 있는 자바스크립트의 핵심 원리

- **원칙**
1. 전역 컨텍스트 생성 후 함수가 호출될 때마다 함수 컨텍스트가 생성된다.
2. 컨텍스트 생성 시 변수 객체, 스코프 체인, this가 생성된다.
3. 컨텍스트 생성 후 함수에 사용되는 변수들은 변수 객체 안에서 값을 고, 없다면 스코프 체인을 따라 올라가며 찾는다.
4. 함수 실행이 마무리되면 해당 컨텍스트는 사라지고(클로저 제외), 페이지가 종료되면 전역 컨텍스트가 사라진다.

**Reply by ez1n** [reply]
https://www.zerocho.com/category/JavaScript/post/5741d96d094da4986bc950a0

**Comment by skyhanull** [comment]
실행 컨텍스트는 자바스크립트 코드가 실행되는 환경이다. 

실행 컨텍스트(execution context)는 실행할 코드에 제공할 환경 정보들을
모아놓은 객체로, 자바스크립트의 동적 언어로서의 성격을 가장 잘 파악할 수 있는
개념이다. --> 코어 자바스크립트에 나온 내용

**Comment by smarfy99** [comment]
실행 컨텍스트는 실행할 코드에 제공할 환경 정보들을 모아놓은 객체이다. 즉, JS 코드가 실행되는 환경이다.

JS 엔진이 스크립트를 처음 마주할 때 전역 컨텍스트를 생성하고, 스크립트를 읽으며 내려가면서, 함수호출을 발견할 때마다 함수의 실행 컨텍스트를 콜 스택에 push 한다.

함수 실행 컨텍스트는 함수가 실행될 때!!!! 만들어진다!!!
